Add getting features from database example

fffffatah opened the pull request:
Simple example for getting features from a database. Example includes,

- A custom FeatureDefinitionProvider.
- Getting features using EFCore and SQLite.

fffffatah commented:
@microsoft-github-policy-service agree

zhiyuanliang-ms commented:
/AzurePipeline run

zhiyuanliang-ms commented on examples/GettingFeaturesFromDatabase/Program.cs:
```diff
@@ -0,0 +1,23 @@
+builder.Services.AddFeatureService();
+
+/* Add feature management */
+builder.Services.AddScoped<IFeatureDefinitionProvider, CustomFeatureDefinitionProvider>().AddScopedFeatureManagement();
```
Hi, @fffffatah Maybe this is a stupid question. May I know the reason why CustomFeatureDefinitionProvider is registered as scoped here?

fffffatah replied:
Hi @zhiyuanliang-ms 👋, CustomFeatureDefinitionProvider is scoped here because the life-time of DbContext and the service which is used to read the features from DB are scoped. If I'm not wrong, setting the life time of DbContext as singleton might lead to issues when many concurrent calls are made. Although, DbContext can be registered as singleton and the Feature set can be marked AsNoTracking while querying, but it felt a bit risky.

zhiyuanliang-ms replied:
Thank you!

zhiyuanliang-ms replied:
Thank you!

zhiyuanliang-ms commented on examples/GettingFeaturesFromDatabase/Database/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        services.AddDbContext<SqliteDbContext>(options =>
+        {
+            options.UseSqlite("Data Source=example.db");
```
What's example.db here?

fffffatah replied:
It's the SQLite database file.

zhiyuanliang-ms replied:
Thanks. Now I see the application will create an example.db during initialization.

zhiyuanliang-ms commented:
@fffffatah When I tried to run this application, it gave the result of "localhost page can not be found". I guess this is because there is no database existed and the feature manager will consider the unexisted feature flag as disabled.

Would you mind including a README? ([example 1](https://github.com/microsoft/FeatureManagement-Dotnet/blob/main/examples/BlazorServerApp/README.md) [example 2](https://github.com/microsoft/FeatureManagement-Dotnet/blob/preview/examples/EvaluationDataToApplicationInsights/README.md))

fffffatah commented:
@zhiyuanliang-ms I simplified the example for initial run and added Swagger UI for ease of use. Also, added a README to get started. Sorry about the exception, it happened due to a package mismatch.

zhiyuanliang-ms commented:
/AzurePipeline run

zhiyuanliang-ms commented:
@fffffatah The application works for me now. But It seems like the CI pipeline failed. I will take some to look into it.

Thank you for your interest and support for this repo. Appreciate you spending the time on this PR.
I want to point out the significant value of this PR to us, even if it ultimately might not be merged.

This PR fills a missing piece of examples in this repo. We provides the IFeatureDefinitionProvider interface and this interface provides the extensibility to allow users to implement their own feature definition provider. But we don't have any detailed tutorial or example about how to implement IFeatureDefinitionProvider, which was called out 3 years ago. (Previous issue can be found in #93)

This PR also lets us know how actually user will use our FM lib to work with feature flags from database. The use case of importing feature flag from database is common especially when users are migrating from other third-party feature management libs. This is super valuable for us. Really appreciate that. So even if this PR cannot be merged, it brings a good chance for us to seriously consider how to build an example app of implementing a custom feature defintion provider from a database. 

Another interesting thing is that this PR uses the `AddScopedFeatureManagement` API which we presented in 3.1.0. The very initial idea of this API is to support Blazor server application to use Feature Management. But it turns out this API can be helpful when working with database according to your comment: https://github.com/microsoft/FeatureManagement-Dotnet/pull/389#discussion_r1518834820 I will spend some time to study why database usage requires scoped services.

Thanks again for your effort. @fffffatah This PR is super valuable for us.

zhiyuanliang-ms commented:
@fffffatch 
The Azure pipeline failed and it reported the following error. It seems the dll of sqlite is not compliant.
![image](https://github.com/microsoft/FeatureManagement-Dotnet/assets/141655842/25330aaa-5702-48dc-bc47-41374072314b)

But I saw your first commit can pass the pipeline and you also used sqlite there.

zhiyuanliang-ms commented:
@fffffatah 

I guess `SQL Server` will not pass the Azure pipeline either. I do remember I met some trouble while including SQL stuff in some example projects.
I tried to include login/register function in an example project, which used SQL server to store user info. That PR failed to pass the Azure pipeline. But I am not 100% sure about whether using SQL server will make the PR fail to pass the CI check. 

Even if a PR can not pass the pipeline check, it still can be merged. So, don't be too worried about it.

Comparing to SQL server, I think SQLite may be a better choice, because it is lighter and user can use it directly as long as the nuget package is installed.
